zulip/zulip-mobile
# Sound storage with SQLite

Here's a prototype to fix #4841, storing soundly in SQLite.

This uses expo-sqlite as the library to make SQLite conveniently accessible from our JS code in React Native, and then provides a small bit of code that provides the same interface as the old AsyncStorage library but with a SQLite database always as the backing store, following more or less the plan from https://github.com/zulip/zulip-mobile/issues/4841#issuecomment-908762299 . It includes a migration to bring the user's data over from the legacy AsyncStorage.

I did a quick manual test on Android, and it worked great:
* Start the app, in an existing install that previously had some data (was logged into an account). It came up smoothly, showing data for that account.
* Add some more data, which would now be stored in the new database: add another account.
* Closed the app and relaunched it; it had the new data.

Still to do before this can be merged:
- [ ] Manually test on iOS too.
- [x] This version breaks our Jest suite, because expo-sqlite doesn't properly run there.

  We've been mocking out the legacy AsyncStorage library, so the simplest solution would be to mock out our replacement AsyncStorage. I might do that just to get this unblocked and merged.

  An improved version would be to get expo-sqlite running in Jest, so that we can have tests exercise our AsyncStorage code. That would be especially good for the sake of its migration code, as migrations tend to be under-exercised by manual testing. It's no surprise that there isn't a handy mock for expo-sqlite -- a mock would basically have to simulate SQLite itself. But surely there are SQLite bindings available for Node (which is the JS engine Jest uses)… and given those, it shouldn't be too hard to get expo-sqlite running in Node, and then under Jest.
- [x] I'd like to prototype first what our future app-level migrations will look like on top of this. That doesn't necessarily need to be complete before merging this, but it may make sense to coalesce the system for those with what this has to manage the migration from the legacy AsyncStorage, and in that case it may influence how we'd like this migrations table to look.
- [ ] There's a commit toward the end that downgrades expo-sqlite, to stay before the "unimodules" -> `expo` migration. I'll want to either drop that, if we've done that migration before this is otherwise ready to merge, or squash it with the commit adding expo-sqlite in the first place, if we haven't.
- (And see comments below.)
